File: server/src/models/CodeSession.ts
import mongoose, { Document, Schema, Types } from 'mongoose';

// TypeScript interface for Snapshot
export interface ISnapshot {
  code: string;
  timestamp: Date;
  userId: Types.ObjectId;
}

// TypeScript interface for ChatMessage
export interface IChatMessage {
  userId: Types.ObjectId;
  username: string;
  message: string;
  timestamp: Date;
}

// TypeScript interface for CodeSession
export interface ICodeSession extends Document {
  roomId: Types.ObjectId;
  snapshots: ISnapshot[];
  chatMessages: IChatMessage[];
  createdAt: Date;
  updatedAt: Date;
}

// Mongoose schema for Snapshot
const snapshotSchema = new Schema<ISnapshot>(
  {
    code: {
      type: String,
      required: true,
    },
    timestamp: {
      type: Date,
      default: Date.now,
    },
    userId: {
      type: Schema.Types.ObjectId,
      ref: 'User',
      required: true,
    },
  },
  {
    _id: false, // Don't create separate _id for subdocuments
  }
);

// Mongoose schema for ChatMessage
const chatMessageSchema = new Schema<IChatMessage>(
  {
    userId: {
      type: Schema.Types.ObjectId,
      ref: 'User',
      required: [true, 'User ID is required'],
    },
    username: {
      type: String,
      required: [true, 'Username is required'],
    },
    message: {
      type: String,
      required: [true, 'Message is required'],
    },
    timestamp: {
      type: Date,
      default: Date.now,
    },
  },
  {
    _id: false, // Don't create separate _id for subdocuments
  }
);

// Mongoose schema for CodeSession
const codeSessionSchema = new Schema<ICodeSession>(
  {
    roomId: {
      type: Schema.Types.ObjectId,
      ref: 'Room',
      required: [true, 'Room ID is required'],
    },
    snapshots: {
      type: [snapshotSchema],
      default: [],
    },
    chatMessages: {
      type: [chatMessageSchema],
      default: [],
    },
  },
  {
    timestamps: true, // Automatically adds createdAt and updatedAt
  }
);

// Limit snapshots array to last 50 entries
codeSessionSchema.pre('save', function (next) {
  if (this.snapshots.length > 50) {
    // Keep only the last 50 snapshots
    this.snapshots = this.snapshots.slice(-50);
  }
  next();
});

// Create unique index on roomId to enforce one session per room
codeSessionSchema.index({ roomId: 1 }, { unique: true });

// Export the model
export const CodeSession = mongoose.model<ICodeSession>('CodeSession', codeSessionSchema);

File: server-spring/src/main/java/com/codeeditor/server/model/CodeSession.java
package com.codeeditor.server.model;

import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.AllArgsConstructor;
import org.springframework.data.annotation.Id;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.mongodb.core.index.Indexed;
import org.springframework.data.mongodb.core.mapping.Document;

import java.time.Instant;
import java.util.ArrayList;
import java.util.List;
import org.springframework.data.mongodb.core.mapping.Field;
import org.springframework.data.mongodb.core.mapping.FieldType;

@Data
@NoArgsConstructor
@AllArgsConstructor
@Document(collection = "codesessions")
public class CodeSession {

    @Id
    private String id;

    @Indexed(unique = true)
    @Field(targetType = FieldType.OBJECT_ID)
    private String roomId; // References Room._id

    private List<Snapshot> snapshots = new ArrayList<>();

    private List<ChatMessage> chatMessages = new ArrayList<>();

    @CreatedDate
    private Instant createdAt;

    @LastModifiedDate
    private Instant updatedAt;

    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    public static class Snapshot {
        private String code;
        private Instant timestamp = Instant.now();
        @Field(targetType = FieldType.OBJECT_ID)
        private String userId;
    }

    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    public static class ChatMessage {
        @Field(targetType = FieldType.OBJECT_ID)
        private String userId;
        private String username;
        private String message;
        private Instant timestamp = Instant.now();
    }

    /**
     * Limit snapshots to last 50 entries (mirrors Mongoose pre-save hook)
     */
    public void trimSnapshots() {
        if (this.snapshots.size() > 50) {
            this.snapshots = new ArrayList<>(this.snapshots.subList(this.snapshots.size() - 50, this.snapshots.size()));
        }
    }
}
